src/dimension_range.c

```diff
@@ -0,0 +1,1127 @@
+{
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	FormData_dimension_range *fd = (FormData_dimension_range *) data;
+
+	Datum values[Natts_dimension_range] = { 0 };
+	bool isnull[Natts_dimension_range] = { 0 };
+	bool doReplace[Natts_dimension_range] = { 0 };
+
+	values[AttrNumberGetAttrOffset(Anum_dimension_range_range_start)] =
+		Int64GetDatum(fd->range_start);
+	doReplace[AttrNumberGetAttrOffset(Anum_dimension_range_range_start)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_dimension_range_range_end)] = Int64GetDatum(fd->range_end);
+	doReplace[AttrNumberGetAttrOffset(Anum_dimension_range_range_end)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_dimension_range_valid)] = BoolGetDatum(fd->valid);
+	doReplace[AttrNumberGetAttrOffset(Anum_dimension_range_valid)] = true;
+
+	HeapTuple new_tuple =
+		heap_modify_tuple(tuple, ts_scanner_get_tupledesc(ti), values, isnull, doReplace);
+
+	ts_catalog_update(ti->scanrel, new_tuple);
+
+	heap_freetuple(new_tuple);
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+static int
+column_dimension_range_scan_internal(ScanKeyData *scankey, int nkeys, tuple_found_func tuple_found,
+									 void *data, int limit, int dimension_index, LOCKMODE lockmode,
+									 MemoryContext mctx)
+{
+	Catalog *catalog = ts_catalog_get();
+	ScannerCtx scanctx = {
+		.table = catalog_get_table_id(catalog, DIMENSION_RANGE),
+		.index = catalog_get_index(catalog, DIMENSION_RANGE, dimension_index),
+		.nkeys = nkeys,
+		.limit = limit,
+		.scankey = scankey,
+		.data = data,
+		.tuple_found = tuple_found,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+		.result_mctx = mctx,
+	};
+
+	return ts_scanner_scan(&scanctx);
+}
+
+int
+ts_dimension_range_update_by_id(int32 dimension_range_id, FormData_dimension_range *fd_range)
+{
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_dimension_range_id_idx_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(dimension_range_id));
+
+	return column_dimension_range_scan_internal(scankey,
+												1,
+												dimension_range_tuple_update,
+												fd_range,
+												1,
+												DIMENSION_RANGE_ID_IDX,
+												RowExclusiveLock,
+												CurrentMemoryContext);
+}
+
+static void
+ts_column_dimension_range_validate(Form_dimension_range info, const Oid hypertable_relid,
+								   bool if_not_exists)
+{
+	HeapTuple tuple;
+	Datum datum;
+	bool isnull;
+
+	/* Check that the column exists and has not been dropped */
+	tuple = SearchSysCacheAttName(hypertable_relid, NameStr(info->column_name));
+
+	if (!HeapTupleIsValid(tuple))
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_COLUMN),
+				 errmsg("column \"%s\" does not exist", NameStr(info->column_name))));
+
+	datum = SysCacheGetAttr(ATTNAME, tuple, Anum_pg_attribute_atttypid, &isnull);
+	Assert(!isnull);
+
+	info->column_type = DatumGetObjectId(datum);
```

Comment:
I am wondering if it is better to get the type from pg_catalog when we read out the dimension_range row instead of storing it in the dimension range.

Reply:
Often we have a pattern of wrapping the FormData in a larger struct that can cache such info, e.g.:

```c
typedef struct DimensionRange {
    FormData_dimension_range fd;
    Oid column_type;
} DimensionRange;

Reply:
Yeah, removed the column type now from the catalog.